x-pack/plugin/sql/src/main/antlr/SqlBase.g4

```diff
@@ -197,25 +207,55 @@ valueExpression
     ;
 
 primaryExpression
-    : CAST '(' expression AS dataType ')'                                            #cast
-    | EXTRACT '(' field=identifier FROM valueExpression ')'                          #extract
+    : castExpression                                                                 #cast
+    | extractExpression                                                              #extract
     | constant                                                                       #constantDefault
     | ASTERISK                                                                       #star
     | (qualifiedName DOT)? ASTERISK                                                  #star
-    | identifier '(' (setQuantifier? expression (',' expression)*)? ')'              #functionCall
+    | functionExpression                                                             #function
     | '(' query ')'                                                                  #subqueryExpression
     | identifier                                                                     #columnReference
     | qualifiedName                                                                  #dereference
     | '(' expression ')'                                                             #parenthesizedExpression
     ;
 
+castExpression
+    : castTemplate                                                                   
+    | FUNCTION_ESC castTemplate ESC_END                                              
+    ;
+    
+castTemplate
+    : CAST '(' expression AS dataType ')'
+    ;
+    
+extractExpression
+    : extractTemplate
+    | FUNCTION_ESC extractTemplate ESC_END
+    ;
+    
+extractTemplate
```

Comment:
Why call these `xxTemplate`?

Reply:
Out of names to use (which need to be unique between rules _and_ callbacks).